### ISSUE TITLE
`C-h f` for some functions link to the loading file instead of their source file

### ISSUE THREAD

**Issue by aavanian** [raxod502/straight.el#1022]
## What's wrong

`C-h f` for some functions link to my `init.e` instead of the actual source file when `straight` is configured to work with `use-package`

## Directions to reproduce

`early-init.el`:

```emacs-lisp
(setq package-enable-at-startup nil)
```

`init.el`

```emacs-lisp

;; block 1
(defvar b,ootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 6))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
(eval-when-compile
  (straight-use-package 'use-package))
(setq straight-use-package-by-default t)

;; block2
;; (eval-when-compile
;;   (add-to-list 'load-path (concat user-emacs-directory "straight/repos/use-package"))
;;   (add-to-list 'load-path (concat user-emacs-directory "straight/repos/consult"))
;;   (add-to-list 'load-path (concat user-emacs-directory "straight/repos/compat"))
;;   (require 'use-package))

;;(straight-use-package 'consult)  ; block 3
(use-package consult)                    ; block 4
```

With the config above:
  - block 1 + block 4: `describe-function` on `consult-multi-occur` (for example) shows `cosult-multi-occur is an autoloaded interactive native-compiled Lisp
function in ‘../../../init.el’.`
  - block 1 + block 3: works correctly (shows `consult-multi-occur is an interactive Lisp closure in ‘consult.el’.`)
  - block 2 + block 4: works correctly (same as above)

Notes
* It started to happen after I upgraded emacs (I was on a few months old build of master without native-comp, now on the 29 branch with native-comp.
* the problem doesn't happen for all modules/functions, it seems it's only autoloaded functions (just an intuition after checking several functions from several packages: consult, embark, marginalia, org-roam)

<details open><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :user-dir "/Users/aavanian/.emacs-profiles/test_consult")
```
</details>

- Test run at: `2022-12-06 14:00:37`
- system-type: `darwin`
- straight-version: `prerelease (HEAD -> master, origin/master, origin/HEAD) 3eca39d 2022-11-25`
- emacs-version: `GNU Emacs 29.0.60 (build 1, aarch64-apple-darwin22.1.0, NS appkit-2299.00 Version 13.0.1 (Build 22A400))
 of 2022-12-05`

<details><summary>Output</summary>

```emacs-lisp
Test run with version: prerelease (HEAD -> master, origin/master, origin/HEAD) 3eca39d 2022-11-25
Packages:
"straight"                n/a                  master 3eca39d 2022-11-25
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master ed9e52d5 2022-12-05
"gnu-elpa-mirror"         n/a                  master a6d1865 2022-12-01
"nongnu-elpa"             n/a                  main ddfb68cf9f 2022-12-03
"el-get"                  melpa                master caf80e20 2022-11-24
"emacsmirror-mirror"      n/a                  master cae49f7 2022-11-22
"use-package"             melpa                master 9078709 2022-11-27
"bind-key"                melpa                master 9078709 2022-11-27
"consult"                 melpa                main df87710 2022-12-05
"compat"                  gnu-elpa-mirror      master 7ca7d30 2022-11-23
```
</details>

**Comment by progfolio** [comment]
Is your init directory symlinked?

**Comment by progfolio** [comment]
See: https://github.com/radian-software/straight.el/issues/944

**Comment by progfolio** [comment]
> there's a bug open upstream

Yes, but it has not seen any activity in quite some time.
It will not get fixed if more people don't chime in on emacs-devel.

> straight has a work-around in order to function but doesn't fix the path in the help?

There's an open pr at #945, I have not tested it, so I'm not sure if it will fix the problem for you.
Ideally the problem would be fixed upstream first (because it may be a more subtle bug than the solution proposed in #945 will solve). The Emacs developers should have some insight to how to fix it.

> with --init-dir option

Do you mean Emacs 29's `--init-directory` command line option?

**Comment by progfolio** [comment]
please try evaluating the following test case and sharing the results of the `*straight-but-report-process*` buffer when it finishes:


<details open><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :user-dir "straight.1022"
  :post-bootstrap 
  (straight-use-package 'consult)
  (message "%s"
           (describe-function #'consult-multi-occur)))
```
</details>

- Test run at: `2022-12-06 02:50:47`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> develop, origin/develop) b89a0f7 2022-11-26`
- emacs-version: `GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.35, cairo version 1.17.6)
 of 2022-12-02`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) b89a0f7 2022-11-26
Cloning consult...
Cloning consult...done
Building consult...
Building consult → Cloning compat...
Building consult → Cloning compat...done
Building consult → Building compat...
Building consult → Building compat...done
Building consult...
Building consult...done

Type M-x help-quit in help window to delete it
consult-multi-occur is an autoloaded interactive Lisp function in
‘consult.el’.

(consult-multi-occur BUFS REGEXP &optional NLINES)

Improved version of ‘multi-occur’ based on ‘completing-read-multiple’.

See ‘multi-occur’ for the meaning of the arguments BUFS, REGEXP and NLINES.


Packages:
"straight"                n/a                  develop b89a0f7 2022-11-26
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master ed9e52d5 2022-12-05
"gnu-elpa-mirror"         n/a                  master a6d1865 2022-12-01
"nongnu-elpa"             n/a                  main ddfb68cf9f 2022-12-03
"el-get"                  melpa                master caf80e20 2022-11-24
"emacsmirror-mirror"      n/a                  master cae49f7 2022-11-22
"consult"                 melpa                main df87710 2022-12-05
"compat"                  gnu-elpa-mirror      master 7ca7d30 2022-11-23


```
</details>

**Comment by progfolio** [comment]
please try the reproduction recipe in the emacs devel thread here and let me know if you are able to reproduce the same issue:

https://lists.gnu.org/r/bug-gnu-emacs/2021-12/msg00376.html

**Comment by Sergey-Makarov** [comment]
Hey guys, I have also faced this problem and have some observations.
TL;DR: setting `straight-cache-autoloads` to `nil` fixes the problem for me.

# My setup:

-   OS: Mac OS
-   Emacs: d12frosted/emacs-plus/emacs-plus@29: stable 29.4

I run emacs with the following command:

    emacs --init-directory=~/dotfiles/emacs_configs/custom.emacs/



# Reproduction steps

Minimal steps to reproduce the issue:

1.  in the init-directory, create init.el file
2.  Place the following  code there
    
        ;;; -*- lexical-binding: t; -*-
        
        ;; straight setup
        (setq straight-use-package-by-default t)
        (defvar bootstrap-version)
        (let ((bootstrap-file
               (expand-file-name
                "straight/repos/straight.el/bootstrap.el"
                (or (bound-and-true-p straight-base-dir)
                    user-emacs-directory)))
              (bootstrap-version 7))
          (unless (file-exists-p bootstrap-file)
            (with-current-buffer
                (url-retrieve-synchronously
                 "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
                 'silent 'inhibit-cookies)
              (goto-char (point-max))
              (eval-print-last-sexp)))
          (load bootstrap-file nil 'nomessage))
        (straight-use-package 'use-package)
        
        ;; package setup
        (use-package smartparens :demand t) ;; demand t is a default, apparently
        ;; macroexpands to essentially (progn (straight-use-package 'smartparens) (require 'smartparens))
        
        ;;testing
        (describe-function 'smartparens-mode)
        (with-current-buffer "*Help*" (rename-uniquely)) ;; rename it to *Help*<2> buffer
        (add-hook 'emacs-startup-hook (lambda ()
                                        (describe-function 'smartparens-mode)))
3.  You will see the following in the two help buffers:
    `*Help*<2>` (during init phase)
    
    > smartparens-mode is an autoloaded, interactive and natively compiled
    > function defined in smartparens.el.
    
    `*Help*` (after init phase)
    
    > smartparens-mode is an autoloaded, interactive and natively compiled
    > function defined in `../../../init.el`.

Let's also check `load-history` variable.
After Emacs is loaded, call `C-h v` `load-history` (note: newest history comes first) to see:

> &#x2026;
> (`/Users/sergey/dotfiles/emacs_configs/custom_plain.emacs/init.el`
> &#x2026;
> (defun . smartparens-mode)
> &#x2026;
> (`/Users/sergey/dotfiles/emacs_configs/custom_plain.emacs/straight/build/smartparens/smartparens.elc`
> &#x2026;
> (defun . smartparens-mode)

That's why `smartparens-mode` is being redefined and messes up symbol's help info.

In comparison, this is how the `load-history` looks like if we add the following line to the init.el file:

    ...
    (straight-use-package 'use-package)
    (setq straight-cache-autoloads nil) ;; this line
    ...

(I also deleted all generated files before relaunching emacs).
`load-history`:

> &#x2026;
>  (`/Users/sergey/dotfiles/emacs_configs/custom_plain.emacs/straight/build/smartparens/smartparens.elc`
>  &#x2026;
>   (defun . smartparens-mode)
> &#x2026;
>  (`/Users/sergey/dotfiles/emacs_configs/custom_plain.emacs/straight/build/smartparens/smartparens-autoloads.el`
>  &#x2026;
>   (defun . smartparens-mode)
> &#x2026;

Now, `C-h f smartparens-mode` shows exactly what it should.



# Conclusions

As I understand it, this problem happens to all autoloaded symbols whose packages are `require`'d during Emacs initialization.
Straight seems to autoload all packages at once after the init.el file is executed, which causes the problem.
So, if the user wants to require the package during the initialization (and they might not even be aware of that, because packages might be required by other packages), they have to do either

1.  set `straight-cache-autoloads` to nil
2.  disable straight autoloading the packages at problem by using `:build (:not autoloads)`

Maybe it's technically possible to make Straight detect such misconfigurations automatically and issue a warning. Or, ideally, not to autoload symbols that have been fully loaded already..

**Comment by raxod502** [comment]
Hmmm. The autoloads loader from `straight.el` is supposed to hack `load-history` to render correctly, even with caching enabled. I guess there's a bug there.

**Comment by raxod502** [comment]
I checked on this. With the reproduction above, things seem working now with the latest version of `straight.el` (note I did recently fix an issue with autoloads caching in #1213, though it's not clear to me why that would have resolved this).

![Image](https://github.com/user-attachments/assets/a02c7817-c007-4178-8877-538ff6eff192)

This is Emacs 30.1 built from source, with autoloads caching enabled (as default). Restarting Emacs (which forces reading from cache) has the same result as above.

Here is the snippet from `load-history` showing init-file loading:

![Image](https://github.com/user-attachments/assets/ca733fba-4013-4a84-b4ed-03c29882e7ae)

There is an entry for defining `smartparens-mode` from the autoloads:

![Image](https://github.com/user-attachments/assets/c824db37-d4b8-4815-9011-4b987768b96f)

And then later (so, earlier in the list), one defining it from the actual library:

![Image](https://github.com/user-attachments/assets/12bf9f3a-2419-496f-a539-4d9f4f5e6af0)

Can anyone still reproduce the issue now that #945 is merged and with the latest version of `straight.el`?
